accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,665 @@
+
+C# in particular just maps calling conventions specified in a function pointer to `System.Runtime.InteropServices.CallConv*` types as metadata. These CallConv* types must live in the same assembly as defines `System.Object` (i.e. `System.Private.Corelib.dll`).
+
+### Normalization of calling conventions
+Calling conventions can be specified in one of two ways in metadata:
+- The older "CallKind" enum (which has issues with not being large enough to hold the newer calling conventions).
+- The newer approach of using modopts on the return parameter.
+
+In ECMA, "CallKind" is used with "CallConv":"
+```
+callConv :'instance' callConv 
+ | 'explicit' callConv 
+ | callKind 
+ ; 
+callKind : /* EMPTY */ 
+ | 'default' 
+ | 'vararg' 
+ | 'unmanaged' 'cdecl' 
+ | 'unmanaged' 'stdcall' 
+ | 'unmanaged' 'thiscall' 
+ | 'unmanaged' 'fastcall' 
+```
+and in exposed in various enums and types. Here's the raw values:
+```
+CALLCONV_DEFAULT       = 0x0,  
+
+CALLCONV_VARARG        = 0x5,  
+CALLCONV_FIELD         = 0x6,  
+CALLCONV_LOCAL_SIG     = 0x7,  
+CALLCONV_PROPERTY      = 0x8,  
+CALLCONV_UNMGD         = 0x9,  
+CALLCONV_GENERICINST   = 0xa,  
+CALLCONV_NATIVEVARARG  = 0xb,  
+CALLCONV_MAX           = 0xc,  
+
+CALLCONV_MASK          = 0x0f,  
+CALLCONV_HASTHIS       = 0x20,  
+CALLCONV_EXPLICITTHIS  = 0x40,  
+CALLCONV_GENERIC       = 0x10  
+```
+
+The calling convention is considered managed if the callKind is "default" or "vararg", so the proposed `Type.IsUnmanagedFunctionPointer` does not require looking at the newer modopts (at least currently; it is possible new managed calling conventions are introduced at some point).
+
+**Open issue: currently this design does not provide a way to determine if the managed calling convention is "default" vs "vararg" or "instance" vs. "explicit" (for the "this" parameter to instance methods).** However, we could easily add new methods to expose this (e.g. using two `bool` properties `IsExplicitThis` and `IsVarArg` or add a future-extensible implementation with enums instead of `bool`).
+
+C# only uses the newer modopts approach when necessary; see the [c# spec section](https://learn.microsoft.com/dotnet/csharp/language-reference/proposals/csharp-9.0/function-pointers#metadata-representation-of-calling-conventions). However, we should assume any valid encoding.
+
+When returning the calling conventions through `Type.GetFunctionPointerCallingConventions()` any calling convention encoded using "CallKind" will return the appropriate CallConv* type, just like it was added through a modopt. This normalization avoids having to expose the "CallKind" in the API and makes it simpler for callers since there is just a single method to return the calling conventions no matter how they were encoded. The `Type.IsUnmanagedFunctionPointer` property was added since there are not CallConv* types for mananged that would otherwise have helped determine this.
```

Comment:
```suggestion
When returning the calling conventions through `Type.GetFunctionPointerCallingConventions()` any calling convention encoded using "CallKind" will return the appropriate CallConv* type, just like it was added through a modopt. This normalization avoids having to expose the "CallKind" in the API and makes it simpler for callers since there is just a single method to return the calling conventions no matter how they were encoded. The `Type.IsUnmanagedFunctionPointer` property was added since there are not CallConv* types for managed that would otherwise have helped determine this.
```